browser/android/youtube_script_injector/brave_youtube_script_injector_native_helper_browsertest.cc

```diff
@@ -0,0 +1,306 @@
+    frame->ExecuteJavaScriptForTests(script, base::NullCallback(),
+                                     content::ISOLATED_WORLD_ID_GLOBAL);
+  }
+
+  bool WaitForJsResult(content::WebContents* web_contents,
+                       const std::string& script) {
+    bool fulfilled = false;
+    TestPredicateWaiter waiter(base::BindRepeating(
+        [](content::WebContents* web_contents, const std::string& script,
+           bool* fulfilled) {
+          bool result = content::EvalJs(web_contents, script).ExtractBool();
+          if (result) {
+            *fulfilled = true;
+          }
+          return result;
+        },
+        web_contents, script, &fulfilled));
+    waiter.Wait();
+    return fulfilled;
+  }
+
+  bool IsVideoPlaying() {
+    return content::EvalJs(web_contents(),
+                           "document.querySelector('video.html5-main-video')."
+                           "paused === false")
+        .ExtractBool();
+  }
+
+ protected:
+  // Must use HTTPS because `youtube.com` is in Chromium's HSTS preload list.
+  net::EmbeddedTestServer https_server_;
+
+ private:
+  content::ContentMockCertVerifier mock_cert_verifier_;
+};
+
+IN_PROC_BROWSER_TEST_F(BraveYouTubeScriptInjectorNativeHelperBrowserTest,
+                       SetFullscreenFromPausedVideo) {
+  const GURL url = https_server_.GetURL("youtube.com", "/yt_fullscreen.html");
+  content::NavigateToURLBlockUntilNavigationsComplete(web_contents(), url, 1,
+                                                      true);
+
+  // Video is not playing initially.
+  ASSERT_FALSE(IsVideoPlaying());
+
+  // The document is not in fullscreen mode.
+  ASSERT_TRUE(
+      content::EvalJs(web_contents(), "document.fullscreenElement === null")
+          .ExtractBool());
+
+  youtube_script_injector::SetFullscreen(web_contents());
+  // Wait for the resize to complete triggered by fullscreen change.
+  content::WaitForResizeComplete(web_contents());
+
+  // Check the video is in fullscreen mode.
+  EXPECT_TRUE(
+      WaitForJsResult(web_contents(), "document.fullscreenElement !== null"));
+
+  // Check the video is playing.
+  EXPECT_TRUE(WaitForJsResult(
+      web_contents(),
+      "document.querySelector('video.html5-main-video').paused === false"));
+}
+
+IN_PROC_BROWSER_TEST_F(BraveYouTubeScriptInjectorNativeHelperBrowserTest,
+                       SetFullscreenFromPlayingVideo) {
+  const GURL url = https_server_.GetURL("youtube.com", "/yt_fullscreen.html");
+  content::NavigateToURLBlockUntilNavigationsComplete(web_contents(), url, 1,
+                                                      true);
+
+  // The document is not in fullscreen mode.
+  ASSERT_TRUE(
+      content::EvalJs(web_contents(), "document.fullscreenElement === null")
+          .ExtractBool());
+
+  // Simulate a playing video before entering fullscreen.
+  ASSERT_TRUE(content::ExecJs(
+      web_contents(),
+      "document.querySelector('video.html5-main-video').play()"));
```

Comment:
does this start playing synchronously in the page or do we need to wait for it to start playing?

Reply:
Oh right! I was not aware but `play()` was actually returning a promise. It didn't compromize the test result but it's best to wait that the promise is resolved.
Addressed by commit 👉 [`597f479` (#29431)](https://github.com/brave/brave-core/pull/29431/commits/597f479524a0dffd9a08248028d951f135d704f8)